Panic during queries with MariaDB

### Rust Version

1.91.1

### Diesel Version

2.3.4

### Diesel Features

"chrono", "r2d2", "numeric"

### Database Version

MariaDB v11 or v12 (Connnector/C v3.4.5 and newer tested)

### Operating System Version

Debian Trixie

### What happened?

Wanted to test out the new fix for MariaDB in the v3.3.4 release of Diesel with Vaultwarden.
Migrations worked fine, but creating an account or logging in is broken.
Tried to use v2.3.3 of Diesel together with the Connector/C v3.4.5 and newer, all works fine (except the newer client lib versions crash with specific conditions of course)

```
[panic][ERROR] thread 'rocket-worker-thread' panicked at 'Libmysqlclient reported a larger size for a fixed size buffer without setting the truncated flag. 
This is a bug somewhere. Please open an issue with reproduction steps at https://github.com/diesel-rs/diesel/issues/new 
Length: 48, Capacity: 48, Type: MYSQL_TYPE_DATETIME': /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.4/src/mysql/connection/bind.rs:430
   0: vaultwarden::init_logging::{{closure}}
             at ./app/src/main.rs:399:25
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/alloc/src/boxed.rs:1999:9
   2: std::panicking::panic_with_hook
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/panicking.rs:842:13
   3: std::panicking::panic_handler::{{closure}}
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/panicking.rs:707:13
   4: std::sys::backtrace::__rust_end_short_backtrace
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/sys/backtrace.rs:174:18
   5: __rustc::rust_begin_unwind
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/panicking.rs:698:5
   6: core::panicking::panic_fmt
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/core/src/panicking.rs:75:14
   7: diesel::mysql::connection::bind::BindData::value
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.4/src/mysql/connection/bind.rs:430:17
   8: <diesel::mysql::connection::stmt::iterator::MysqlField as diesel::row::Field<diesel::mysql::backend::Mysql>>::value
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.4/src/mysql/connection/stmt/iterator.rs:227:84
   9: <vaultwarden::db::multi_connection_impl::row::MultiField as diesel::row::Field<vaultwarden::db::multi_connection_impl::backend::MultiBackend>>::value
             at ./app/src/db/mod.rs:47:10
  10: <T as diesel::deserialize::FromStaticSqlRow<ST,DB>>::build_from_row
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.4/src/deserialize.rs:574:36
  11: diesel::type_impls::tuples::<impl diesel::deserialize::FromStaticSqlRow<(ST1,ST2,ST3,ST4,ST5,ST6,ST7,ST8,ST9,ST10,ST11,ST12,ST13,ST14,ST15,ST16,ST17,ST18,ST19,ST20,ST21,ST22,ST23,ST24,ST25,ST26,ST27,ST28,ST29,ST30,ST),__DB> for (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,T21,T22,T23,T24,T25,T26,T27,T28,T29,T30,T)>::build_from_row
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.4/src/deserialize.rs:559:19
  12: diesel::query_dsl::load_dsl::private::LoadIter<U,C,ST,DB>::map_row
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.4/src/deserialize.rs:559:19
  13: <diesel::query_dsl::load_dsl::private::LoadIter<U,C,ST,DB> as core::iter::traits::iterator::Iterator>::next
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.4/src/query_dsl/load_dsl.rs:148:13
  14: diesel::query_dsl::RunQueryDsl::get_result
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.4/src/query_dsl/mod.rs:1778:41
  15: diesel::query_dsl::RunQueryDsl::first
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.4/src/query_dsl/mod.rs:1835:43
  16: vaultwarden::db::models::user::User::find_by_mail::{{closure}}::{{closure}}
             at ./app/src/db/models/user.rs:375:18
  17: vaultwarden::db::DbConn::run::{{closure}}::{{closure}}
             at ./app/src/db/mod.rs:317:45
  18: tokio::runtime::context::runtime_mt::exit_runtime
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/context/runtime_mt.rs:35:5
  19: tokio::runtime::scheduler::multi_thread::worker::block_in_place
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/scheduler/multi_thread/worker.rs:448:9
  20: tokio::runtime::scheduler::block_in_place::block_in_place
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/scheduler/block_in_place.rs:8:5
  21: tokio::task::blocking::block_in_place
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/task/blocking.rs:78:9
  22: vaultwarden::db::DbConn::run::{{closure}}
             at ./app/src/db/mod.rs:317:9
  23: vaultwarden::db::models::user::User::find_by_mail::{{closure}}
             at ./app/src/db/mod.rs:324:39
  24: vaultwarden::api::core::accounts::_prelogin::{{closure}}
             at ./app/src/api/core/accounts.rs:1239:96
  25: vaultwarden::api::identity::prelogin::{{closure}}
             at ./app/src/api/identity.rs:878:27
  26: vaultwarden::api::identity::prelogin::into_info::monomorphized_function::{{closure}}
             at ./app/src/api/identity.rs:877:1
  27: <core::pin::Pin<P> as core::future::future::Future>::poll
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/core/src/future/future.rs:133:9
  28: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::future::future::Future>::poll
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/core/src/panic/unwind_safe.rs:299:9
  29: <futures_util::future::future::catch_unwind::CatchUnwind<Fut> as core::future::future::Future>::poll::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/futures-util-0.3.31/src/future/future/catch_unwind.rs:37:44
  30: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/core/src/panic/unwind_safe.rs:274:9
  31: std::panicking::catch_unwind::do_call
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/panicking.rs:590:40
  32: __rust_try
  33: std::panic::catch_unwind
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/panicking.rs:553:19
  34: <futures_util::future::future::catch_unwind::CatchUnwind<Fut> as core::future::future::Future>::poll
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/futures-util-0.3.31/src/future/future/catch_unwind.rs:37:9
  35: rocket::server::handle::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rocket-0.5.1/src/server.rs:57:10
  36: rocket::server::<impl rocket::rkt::Rocket<rocket::phase::Orbit>>::route::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rocket-0.5.1/src/server.rs:335:80
  37: rocket::server::<impl rocket::rkt::Rocket<rocket::phase::Orbit>>::route_and_process::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rocket-0.5.1/src/server.rs:288:60
  38: rocket::server::<impl rocket::rkt::Rocket<rocket::phase::Orbit>>::dispatch::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rocket-0.5.1/src/server.rs:262:66
  39: rocket::server::hyper_service_fn::{{closure}}::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rocket-0.5.1/src/server.rs:90:71
  40: <core::pin::Pin<P> as core::future::future::Future>::poll
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/core/src/future/future.rs:133:9
  41: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/core.rs:365:24
  42: tokio::runtime::task::core::Core<T,S>::poll
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/loom/std/unsafe_cell.rs:16:9
  43: tokio::runtime::task::harness::poll_future::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/harness.rs:535:30
  44: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/core/src/panic/unwind_safe.rs:274:9
  45: std::panicking::catch_unwind::do_call
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/panicking.rs:590:40
  46: __rust_try
  47: std::panic::catch_unwind
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/panicking.rs:553:19
  48: tokio::runtime::task::harness::poll_future
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/harness.rs:523:18
  49: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/harness.rs:210:27
  50: tokio::runtime::task::harness::Harness<T,S>::poll
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/harness.rs:155:20
  51: tokio::runtime::task::raw::poll
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/raw.rs:325:13
  52: tokio::runtime::task::raw::RawTask::poll
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/raw.rs:255:18
  53: tokio::runtime::task::LocalNotified<S>::run
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/mod.rs:509:13
  54: tokio::runtime::scheduler::multi_thread::worker::Context::run_task::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/scheduler/multi_thread/worker.rs:600:18
  55: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/task/coop/mod.rs:167:5
  56: tokio::runtime::scheduler::multi_thread::worker::Context::run
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/scheduler/multi_thread/worker.rs:539:29
  57: tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/scheduler/multi_thread/worker.rs:504:24
  58: tokio::runtime::context::scoped::Scoped<T>::set
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/context/scoped.rs:40:9
  59: tokio::runtime::context::set_scheduler::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/context.rs:176:38
  60: std::thread::local::LocalKey<T>::try_with
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/thread/local.rs:315:12
  61: std::thread::local::LocalKey<T>::with
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/thread/local.rs:279:20
  62: tokio::runtime::context::set_scheduler
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/context.rs:176:17
  63: tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/scheduler/multi_thread/worker.rs:499:9
  64: tokio::runtime::context::runtime::enter_runtime
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/context/runtime.rs:65:16
  65: tokio::runtime::scheduler::multi_thread::worker::run
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/scheduler/multi_thread/worker.rs:491:5
  66: tokio::runtime::scheduler::multi_thread::worker::Launch::launch::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/scheduler/multi_thread/worker.rs:457:45
  67: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/blocking/task.rs:42:21
  68: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/core.rs:365:24
  69: tokio::runtime::task::core::Core<T,S>::poll
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/loom/std/unsafe_cell.rs:16:9
  70: tokio::runtime::task::harness::poll_future::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/harness.rs:535:30
  71: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/core/src/panic/unwind_safe.rs:274:9
  72: std::panicking::catch_unwind::do_call
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/panicking.rs:590:40
  73: __rust_try
  74: std::panic::catch_unwind
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/panicking.rs:553:19
  75: tokio::runtime::task::harness::poll_future
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/harness.rs:523:18
  76: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/harness.rs:210:27
  77: tokio::runtime::task::harness::Harness<T,S>::poll
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/harness.rs:155:20
  78: tokio::runtime::task::raw::poll
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/raw.rs:325:13
  79: tokio::runtime::task::raw::RawTask::poll
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/raw.rs:255:18
  80: tokio::runtime::task::UnownedTask<S>::run
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/task/mod.rs:546:13
  81: tokio::runtime::blocking::pool::Task::run
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/blocking/pool.rs:161:19
  82: tokio::runtime::blocking::pool::Inner::run
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/blocking/pool.rs:516:22
  83: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at ./root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.48.0/src/runtime/blocking/pool.rs:474:47
  84: std::sys::backtrace::__rust_begin_short_backtrace
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/sys/backtrace.rs:158:18
  85: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/thread/mod.rs:559:17
  86: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/core/src/panic/unwind_safe.rs:274:9
  87: std::panicking::catch_unwind::do_call
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/panicking.rs:590:40
  88: __rust_try
  89: std::thread::Builder::spawn_unchecked_::{{closure}}
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/panicking.rs:553:19
  90: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/core/src/ops/function.rs:250:5
  91: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/alloc/src/boxed.rs:1985:9
  92: std::sys::thread::unix::Thread::new::thread_start
             at ./rustc/ed61e7d7e242494fb7057f2657300d9e77bb4fcb/library/std/src/sys/thread/unix.rs:126:17
  93: <unknown>
  94: __clone
```

### What did you expect to happen?

No panic

### Additional details

Didn't tried to create an example yet, as i have no time right now. Maybe it's easy reproducible.

### Steps to reproduce

```rust
1. Pull Vaultwarden repo (github.com/dani-garcia/vaultwarden)
2. Update all the crates
3. run `CARGO_PROFILE=dev docker/bake.sh debian-amd64`
4. Start the container, something like this

docker run --rm -it \
  -e DISABLE_ADMIN_TOKEN=true \
  -e I_REALLY_WANT_VOLATILE_STORAGE=true \
  -e LOG_LEVEL=debug \
  -e ROCKET_PORT=8080 \
  -e DOMAIN='http://localhost:8080' \
  -e DATABASE_URL="mysql://user:pass@127.0.0.1:3306/vw" \
  --network=host \
  vaultwarden/server:testing-amd64
```

### Compile time error

None

### Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) and the [discussion forum](https://github.com/diesel-rs/diesel/discussions) for similar possible closed issues.

- [x] This issue can be reproduced on Rust's stable channel. (Please submit the issue in the Rust issue tracker instead if that's not the case)
- [x] This issue can be reproduced without requiring a third party crate

--- a/diesel/src/mysql/connection/bind.rs
+++ b/diesel/src/mysql/connection/bind.rs
@@ -427,11 +427,13 @@ impl BindData {
             // We assert that we don't read more than capacity bytes below as that's
             // the most important invariant to uphold here
             let length = if let Some(length) = known_buffer_size_for_ffi_type(self.tpe) {
-                debug_assert!(length <= self.length.try_into().expect("Usize is at least 32 bit"),
+                debug_assert!(
+                    length <= self.capacity
+                    && <usize as TryFrom<_>>::try_from(self.length).expect("32bit integer fits in a usize")  <= self.capacity,
                     "Libmysqlclient reported a larger size for a fixed size buffer without setting the truncated flag. \n\
                      This is a bug somewhere. Please open an issue with reproduction steps at \
                      https://github.com/diesel-rs/diesel/issues/new \n\
-                     Length: {length}, Capacity: {}, Type: {:?}", self.capacity, self.tpe
+                     Calculated Length: {length}, Buffer Capacity: {}, Reported Length {}, Type: {:?}", self.capacity, self.length, self.tpe
                 );
                 length
             } else {

